### PR TITLE
Baked data for `icu_datetime`

### PR DESCRIPTION
**Pull request by robertbastian**
#3576
#2945

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

File: components/datetime/src/raw/datetime.rs [hunk 1]
```diff
@@ -297,11 +373,65 @@ impl DateTimeFormatter {
         })
     }
 
-    /// Constructor that takes a selected [`Locale`], reference to a [`DataProvider`] and
-    /// a list of options, then collects all data necessary to format date and time values into the given locale.
     #[inline(never)]
-    pub fn try_new<D>(
-        data_provider: &D,
+    #[cfg(feature = "data")]
+    pub fn try_new(
```
Comment (sffc) [inline components/datetime/src/raw/datetime.rs:378]: Thought: It would be nice if we could avoid duplicating this code but I don't see a way to do it so I think duplicating is the least offensive option

File: components/datetime/src/time_zone.rs [hunk 2]
```diff
@@ -463,11 +463,45 @@ impl TimeZoneFormatter {
         })
     }
 
-    icu_provider::gen_any_buffer_constructors!(
-        locale: include,
-        options: TimeZoneFormatterOptions,
-        error: DateTimeError
-    );
+    /// Load generic non location long format for timezone. For example, Pacific Time.
```
Comment (sffc) [inline components/datetime/src/time_zone.rs:466]:
```suggestion
    /// Load generic non location long format for timezone from compiled data. For example, Pacific Time.
```

File: components/datetime/src/time_zone.rs [hunk 3]
```diff
@@ -463,11 +463,45 @@ impl TimeZoneFormatter {
...
+    #[cfg(feature = "data")]
+    pub fn with_generic_non_location_long(
```
Comment (sffc) [inline components/datetime/src/time_zone.rs:468]: Nit: I think I prefer `insert_` as the prefix because `with_` usually refers to something immutable in Java/C++ (sorry, I know I suggested it)

```suggestion
    pub fn insert_generic_non_location_long(
```

File: components/locid_transform/src/fallback/algorithms.rs [hunk 4]
```diff
@@ -243,7 +244,7 @@ mod tests {
             requires_data: false,
             extension_key: None,
             fallback_supplement: None,
-            expected_language_chain: &["en-US-u-sd-usca", "en-US", "en"],
+            expected_language_chain: &["en-US-u-sd-usca", "en-US", "en-u-sd-usca", "en"],
```
Comment (sffc) [inline components/locid_transform/src/fallback/algorithms.rs:247]: Nit: I would prefer if you put the locid_transform changes into its own PR because it is a standalone behavior change that should be independently trackable. If this is too hard (don't know why it would be), then change this PR title to mention it.

File: components/datetime/src/raw/datetime.rs (hunk 1 above)
Comment (robertbastian) [reply components/datetime/src/raw/datetime.rs:378]: We did decide to have different constructor trees for recursive constructors, this is the result

File: components/datetime/src/time_zone.rs (hunk 3 above)
Comment (robertbastian) [reply components/datetime/src/time_zone.rs:468]: `insert` sounds a lot like a collection

File: components/locid_transform/src/fallback/algorithms.rs (hunk 4 above)
Comment (robertbastian) [reply components/locid_transform/src/fallback/algorithms.rs:247]: #3594

File: components/datetime/src/raw/datetime.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/src/raw/datetime.rs:378]: Yes I understand. We also said that when constructors needed to be duplicated, we needed to ensure good test coverage for both paths. Can you make sure the old code path is covered by sufficient tests?

File: components/datetime/src/time_zone.rs (hunk 3 above)
Comment (Manishearth) [reply components/datetime/src/time_zone.rs:468]: yeah, I prefer `with` (or potentially `include`)
